smarts/sstudio/types.py

```diff
@@ -462,50 +463,85 @@ class MapZone(Zone):
     """The number of lanes from right to left that this zone covers."""
 
     def to_geometry(self, road_network: SumoRoadNetwork) -> Polygon:
-        def resolve_offset(offset, geometry_length, lane_length, buffer_from_ends):
-            if offset == "base" or offset == 0:
-                return buffer_from_ends
+        def resolve_offset(offset, geometry_length, lane_length):
+            if offset == "base":
+                return 0
             # push off of end of lane
             elif offset == "max":
-                return lane_length - geometry_length - buffer_from_ends
+                return lane_length - geometry_length
             elif offset == "random":
-                return random.uniform(
-                    0, lane_length - geometry_length - buffer_from_ends
-                )
+                return random.uniform(0, lane_length - geometry_length)
             else:
                 return float(offset)
 
+        def pick_remaining_shape_after_split(geometry_collection, length, lane):
+            lane_shape = geometry_collection
+            if not isinstance(lane_shape, GeometryCollection):
+                return lane_shape
+
+            # For simplicty, we only deal w/ the == 1 or 2 case
+            if len(lane_shape) == 0 or len(lane_shape) > 2:
+                return None
+
+            if len(lane_shape) == 1:
+                return lane_shape[0]
+
+            expected_bbox_area = lane.getWidth() * length
+            keep_index = 0
+            if (lane_shape[0].minimum_rotated_rectangle.area - expected_bbox_area) < (
+                lane_shape[1].minimum_rotated_rectangle.area - expected_bbox_area
+            ):
+                # 0 is the discard piece, keep the other
+                keep_index = 1
+            lane_shape = lane_shape[keep_index]
+
+            return lane_shape
+
         lane_shapes = []
         edge_id, lane_idx, offset = self.start
         edge = road_network.edge_by_id(edge_id)
+        buffer_from_ends = 1e-6
         for lane_idx in range(lane_idx, lane_idx + self.n_lanes):
             lane = edge.getLanes()[lane_idx]
             lane_length = lane.getLength()
-            geom_length = max(self.length - 1e-6, 1e-6)
+            geom_length = self.length
+
+            if geom_length > lane_length:
+                logging.debug(
+                    f"Geometry is too long={geom_length} with offset={offset} for "
+                    f"lane={lane.getID()}, using length={lane_length} instead"
+                )
+                geom_length = lane_length
 
-            assert lane_length > geom_length  # Geom is too long for lane
             assert geom_length > 0  # Geom length is negative
 
-            lane_shape = SumoRoadNetwork.buffered_lane_or_edge(
+            lane_shape = SumoRoadNetwork._buffered_lane_or_edge(
                 lane, width=lane.getWidth() + 0.3
```

Comment:
It might be best to convert the `0.3` to a padding argument or remove if possible.